### ISSUE TITLE
freezing or unfreezing coins very slow

### ISSUE THREAD

**Issue by zveda** [Electron-Cash/Electron-Cash#2102]
On a wallet with a few hundred coins and a few thousand addresses, freezing all coins can take as much as half an hour, despite no communication with the server. The wallet is also locked during this time. Likely culprit of slowness is the is_mine() calls within the set_frozen_coin_state() function in wallet.py.

**Comment by cculianu** [comment]
You're right that is the culprit.  It makes the algorithm quadratic.  Hmm. I'll see about just removing that check. It's not even really necessary.

**Comment by cculianu** [comment]
Ok so I did some investigation -- it turns out `is_mine` is complexity logN (there N is the number of addresses in the wallet).  I am going to push a commit but it doesn't buy us much.  On a 40000 coin wallet with 1500 addresses, I got a 90ms going down to 45ms without the `is_mine` check.  That's not a huge difference.

I believe your slowdown was from the GUI updates on a huge `Coins` tab.  Re-rendering the `Coins` tab is slow.  You can avoid this by doing the following, go to the `Console` tab and do:

`wallet.set_frozen_coin_state(wallet.get_utxos(), True)`

This freezes all coins in your wallet (pass `False` to unfreeze).  You will see it runs very fast. If you switch to `Console` tab it will not have updated the GUI.  You can hit Ctrl+R (refresh) to update the GUI.  I believe updating the GUI is the culprit.

Anyway I will push a commit to skip the `is_mine` check since it's not even really necessary -- but unfortunately it won't really solve your problem.  Updating the GUI is just... slow.

**Comment by zveda** [comment]
Thanks for that. Can confirm the console trick makes freezing or unfreezing all coins almost instant, down from 15-30 minutes via GUI previously.

**Comment by markblundeberg** [comment]
Just to be clear -- the slowness is proportional to total number of unspent coins in the wallet (i.e., how many entries get placed in the coins tab table)?

**Comment by zveda** [comment]
> Just to be clear -- the slowness is proportional to total number of unspent coins in the wallet (i.e., how many entries get placed in the coins tab table)?

Yes I am quite sure this is right.
